storage/scanner.go

```diff
@@ -179,58 +226,47 @@ func (rs *rangeScanner) scanLoop(clock *hlc.Clock, stopper *util.Stopper) {
 		stats := &storeStats{}
 
 		for {
-			waitInterval := rs.paceInterval(start, time.Now())
-			if log.V(6) {
-				log.Infof("Wait time interval set to %s", waitInterval)
-			}
-			select {
-			case <-time.After(waitInterval):
-				if !stopper.StartTask() {
-					continue
-				}
-				rng := rs.iter.Next()
-				if rng != nil {
-					// Try adding range to all queues.
-					for _, q := range rs.queues {
-						q.MaybeAdd(rng, clock.Now())
-					}
-					stats.RangeCount++
-					ms := rng.stats.GetMVCC()
-					stats.MVCC.Add(&ms)
-				} else {
-					// Otherwise, we're done with the iteration. Reset iteration and start time.
-					rs.iter.Reset()
-					start = time.Now()
-					// Store the most recent scan results in the scanner's stats.
-					atomic.StorePointer(&rs.stats, unsafe.Pointer(stats))
-					stats = &storeStats{}
-					if rs.scanFn != nil {
-						rs.scanFn()
-					}
-					// Increment iteration count.
-					rs.completedScan.L.Lock()
-					rs.count++
-					rs.completedScan.Broadcast()
-					rs.completedScan.L.Unlock()
-					if log.V(6) {
-						log.Infof("reset range scan iteration")
-					}
-				}
-				stopper.FinishTask()
-
-			case rng := <-rs.removed:
-				// Remove range from all queues as applicable.
-				for _, q := range rs.queues {
-					q.MaybeRemove(rng)
+			if rs.ranges.Count() == 0 {
+				// Just wait without processing any range.
+				if rs.waitAndProcess(start, clock, stopper, stats, nil) {
+					break
 				}
-				if log.V(6) {
-					log.Infof("removed range %s", rng)
+			} else {
+				shouldStop := true
+				rs.ranges.Visit(func(rng *Range) bool {
+					shouldStop = rs.waitAndProcess(start, clock, stopper, stats, rng)
```

Comment:
The only time shouldStop would be true is if rng is nil, so I think we should just remove the return value from waitAndProcess and unconditionally return true in the visit function.